### PR TITLE
Make monster::heal actually return amount healed

### PR DESCRIPTION
**Pull request by actual-nh**
#### Summary
Bugfixes "Make monster::heal actually return the amount healed"

#### Purpose of change

It is assumed in multiple places, and stated in monster.h, that the monster::heal function returns the amount healed. It doesn't, but it should.

#### Describe the solution

To get the return value, do hp - old_hp, not maxhp - old_hp.

#### Describe alternatives you've considered

This came about via #17452, which was to prevent heal from returning a negative number; I see no way that it now can, but just in case I have put in a std::max( 0, hp - old_hp ). It may be preferable to not do the max, or to emit an error if the heal amount somehow manages to be negative.

#### Testing

CI and by @eltank; thanks very much!

#### Additional context

See dev discord.

### CODE REVIEW

**Comment by eltank** [comment]
Please change `monster::on_load()` as well, otherwise it will be broken by this change.

```
    if( healed < heal_amount && get_speed_base() < type->speed ) {
        int old_speed = get_speed_base();
        set_speed_base( std::min( get_speed_base() + heal_amount - healed, type->speed ) );
        healed_speed = get_speed_base() - old_speed;
    }
```

We need to give the monster its full speed if they did a full heal.
Also I think the partial speed heal should be proportional to hp healed as a fraction of total hp.
```
    if( get_speed_base() < type->speed ) {
        int old_speed = get_speed_base();
        if( hp >= type->hp ) {
            set_speed_base( type-> speed );
        } else {
            set_speed_base( std::min( old_speed + healed * type->speed / type->hp, type->speed ) );
        }
        healed_speed = get_speed_base() - old_speed;
    }
```

**Comment by actual-nh** [comment]
> Please change `monster::on_load()` as well, otherwise it will be broken by this change.
> 
> ```
>     if( healed < heal_amount && get_speed_base() < type->speed ) {
>         int old_speed = get_speed_base();
>         set_speed_base( std::min( get_speed_base() + heal_amount - healed, type->speed ) );
>         healed_speed = get_speed_base() - old_speed;
>     }
> ```

I'm not sure how this change would break the above. Healed is the amount that was actually healed. If the potential heal amount is more than how much actually got healed, the "extra" goes to increasing speed... rather oddly, I agree.

> We need to give the monster its full speed if they did a full heal.
> Also I think the partial speed heal should be proportional to hp healed as a fraction of total hp.

I agree with both of the above; will do the below, thank you.

> ```
>     if( get_speed_base() < type->speed ) {
>         int old_speed = get_speed_base();
>         if( hp >= type->hp ) {
>             set_speed_base( type-> speed );
>         } else {
>             set_speed_base( std::min( old_speed + healed * type->speed / type->hp, type->speed ) );
>         }
>         healed_speed = get_speed_base() - old_speed;
>     }
> ```

File: src/monster.cpp [hunk 1]
```diff
@@ -1772,7 +1772,7 @@ int monster::heal( const int delta_hp, bool overheal )
     if( hp > maxhp && !overheal ) {
         hp = maxhp;
     }
-    return maxhp - old_hp;
+    return std::max( 0, hp - old_hp );
```
Comment (eltank) [inline src/monster.cpp:1775]: The std::max is not needed. Health can not decrease.
This is due to delta_hp > 0 and hp < maxhp which is checked above.

Comment (actual-nh) [inline src/monster.cpp:1775]:
```suggestion
    return hp - old_hp;
```
That's what I thought, but I'm a bit paranoid at times...

**Comment by eltank** [comment]
> I'm not sure how this change would break the above. Healed is the amount that was actually healed. If the potential heal amount is more than how much actually got healed, the "extra" goes to increasing speed... rather oddly, I agree.

Currently that if condition evaluates to true if the monster was close to max hp. After your change (and without the additional modifications) the condition would evaluate to false when the monster is close to max and regen is capped. That would prevent it from healing speed. But the code is broken anyway when monster hp is low (if condition evaluates to false). The suggested change should fix all that.

**Comment by actual-nh** [comment]
> > I'm not sure how this change would break the above. Healed is the amount that was actually healed. If the potential heal amount is more than how much actually got healed, the "extra" goes to increasing speed... rather oddly, I agree.
> 
> Currently that if condition evaluates to true (kind of by accident) if the monster was close to max hp. After your change (and without the additional modifications) the condition would evaluate to false when the monster is close to max and regen is capped. That would prevent it from healing speed. But the code was broken anyway when monster hp is low (the if condition evaluates to false). The suggested change should fix all that.

Actually, healed will be less than heal_amount if it's capped, meaning it will heal speed. But I agree that the code is confusing and doesn't make much sense, at least considering that there doesn't seem to be any _other_ healing of speed _in_ the reality bubble!

**Comment by actual-nh** [comment]
Should the proportional healing of speed be rounded up?

**Comment by eltank** [comment]
> Should the proportional healing of speed be rounded up?

Probably doesn't matter since they'll likely do a full heal anyway, but you can change to lround or round up or even roll_remainder. This is just the simplest option.

**Comment by actual-nh** [comment]
> > Should the proportional healing of speed be rounded up?
> 
> Probably doesn't matter since they'll likely do a full heal anyway, but you can change to lround or round up or even roll_remainder if you prefer. This is just the simplest option.

Just in case, making sure there's always a bit of healing of speed; also fixing a typo that I'm surprised astyle didn't object to - space between `type->` and `speed`.

**Comment by eltank** [comment]
Probably out of scope for this PR, but regenerating monsters seem to only regenerate speed when they enter the reality bubble (on_load) and not while in the reality bubble (in which case they only regen hp).

Not sure what the implications are. AFAIK the only way monsters can lose base speed is during resurrection.

File: src/monster.cpp [hunk 2]
```diff
@@ -3118,9 +3118,14 @@ void monster::on_load()
...
     const int healed = heal( heal_amount );
     int healed_speed = 0;
-    if( healed < heal_amount && get_speed_base() < type->speed ) {
+    if( get_speed_base() < type->speed ) {
         int old_speed = get_speed_base();
-        set_speed_base( std::min( get_speed_base() + heal_amount - healed, type->speed ) );
+        if( hp >= type->hp ) {
+            set_speed_base( type->speed );
+        } else {
+            set_speed_base( std::min( old_speed + std::ceil( heal_amount * type->speed / type->hp ),
```
Comment (eltank) [inline src/monster.cpp:3126]:
```suggestion
            int speed_delta = std::max( healed * type->speed / type->hp, 1 );
            set_speed_base( std::min( old_speed + speed_delta, type->speed ) );
```

std::ceil works on floating point arguments and returns floating point; you're giving it an int

File: src/monster.cpp [hunk 3]
```diff
@@ -3118,9 +3118,14 @@ void monster::on_load()
     const int heal_amount = roll_remainder( regen * to_turns<int>( dt ) );
     const int healed = heal( heal_amount );
     int healed_speed = 0;
-    if( healed < heal_amount && get_speed_base() < type->speed ) {
+    if( get_speed_base() < type->speed ) {
         int old_speed = get_speed_base();
```
Comment (eltank) [inline src/monster.cpp:3122]:
```suggestion
        const int old_speed = get_speed_base();
```

**Comment by actual-nh** [comment]
> Probably out of scope for this PR, but regenerating monsters seem to only regenerate speed when they enter the reality bubble (on_load) and not while in the reality bubble (in which case they only regen hp).
> 
> Not sure what the implications are. AFAIK the only way monsters can lose base speed is during resurrection, which I think can only happen in the bubble anyway. So when would speed regen on load even trigger?

Scenario: Something manages to revive, but you move and it's no longer in the reality bubble. You come back, it should have regenerated speed as well as hp in the meantime. (Of course, it should also be doing that while in the reality bubble... hmm.)

> The main implication of that is: I don't know how to test that the speed regen change works.
> Hmm, maybe wait for a dissolute to rez, go away then come back later.

Sounds good to me!

Oh. I think I just realized why it's only at times (resurrection or loading) that the monster isn't interacting that the speed is changing. If a monster regenerated sufficient speed to do another action by the end of a turn, but didn't have that speed at the beginning of the turn, where does it fall in the ordering of actions? Admittedly, the same is true of a PC donning a ring of speed/haste/whatever...

**Comment by eltank** [comment]
> If a monster regenerated sufficient speed to do another action by the end of a turn, but didn't have that speed at the beginning of the turn, where does it fall in the ordering of actions?

Effect processing (including regen) happens before it gets its moves (at the beginning of the turn).

File: src/monster.cpp [hunk 4]
```diff
@@ -3118,9 +3118,15 @@ void monster::on_load()
...
-        int old_speed = get_speed_base();
-        set_speed_base( std::min( get_speed_base() + heal_amount - healed, type->speed ) );
+    if( get_speed_base() < type->speed ) {
+        const int old_speed = get_speed_base();
+        if( hp >= type->hp ) {
+            set_speed_base( type->speed );
+        } else {
+            const int speed_delta = std::max( healed * type->speed / type->hp, 1 );
+            set_speed_base( std::min( old_speed + speed_delta, type->speed ) );
+                                      type->speed ) );
```
Comment (eltank) [inline src/monster.cpp:3128]:
```suggestion
            set_speed_base( std::min( old_speed + speed_delta, type->speed ) );
```

Something got messed up here. Let's remove the junk.

Comment (actual-nh) [reply src/monster.cpp:3128]: Github is insisting that there's no change there - doing a pull to my local machine to check.

**Comment by eltank** [comment]
First round of testing. You can see that before this change, dissolute regen was reported as "visibly regenerating", which should be displayed when healing more than 50 points. But dissolutes only have 1 point of regen.

image

After the change, the report is "healing slowly", which is what is expected for this monster:

image

Note that even after regaining all hp, they can barely move after rez if you keep then in the bubble:

image

**Comment by eltank** [comment]
Test 2. Kill a dissolute, wait for it to rez, teleport away, wait 1s and teleport back. With and without this change the outcome is the same. They somehow get a full heal (which seems broken) and recover their speed. So none of that partial heal code is kicking in...

image

image

**Comment by actual-nh** [comment]
> Test 2. Kill a dissolute, wait for it to rez, teleport away, wait 1s and teleport back. With and without this change the outcome is the same. They somehow get a full heal (which seems broken) and recover their speed. So none of that partial heal code is kicking in...

Odd. Did you catch the debug message at the end of on_load()?

**Comment by actual-nh** [comment]
> The problem seems to be that the number of turns since last update is incorrect:
> image
> Expected 1 or 2, but it was 5 million!

I was wondering if it might be something with that. `to_turns<int>( dt )` (or just maybe `const time_duration dt = calendar::turn - last_updated`) would appear not to be behaving correctly.

**Comment by eltank** [comment]
> I was wondering if it might be something with that. `to_turns<int>( dt )` (or just maybe `const time_duration dt = calendar::turn - last_updated`) would appear not to be behaving correctly.

No, that part is working fine.
The `last_updated` value is set correctly in `on_unload` but then it gets lost. When `on_load` executes it is 0.

**Comment by actual-nh** [comment]
> The `last_updated` value is set correctly in `on_unload` but then it gets lost. When `on_load` executes it is 0.

Oh, interesting. (I note that npc::on_unload() doesn't even _try_ to set last_updated?)

**Review by eltank — APPROVED** [review]

**Comment by eltank** [comment]
It looks like that last_update bug is going to take a while to debug. 
In the mean time, I verified that this PR does fix one bug, and the code looks good to me so this should be merge-able.

**Comment by eltank** [comment]
I fixed that other bug, so I repeated test 2. I teleported out after the dissolute got up, waited 15 turns, teleported back, and saw:

image

Partial hp restored and partial speed restored as expected.

**Comment by actual-nh** [comment]
> I fixed that other bug, so I repeated test 2. I teleported out after the dissolute got up, waited 15 turns, teleported back, and saw:
> 
> image
> 
> Partial hp restored and partial speed restored as expected.

Excellent! (What was the bug, BTW? I look forward to a PR to fix it; that will also be very helpful for having morale and anger properly affected by time out of the reality bubble.)

**Comment by eltank** [comment]
> Excellent! (What was the bug, BTW? I look forward to a PR to fix it; that will also be very helpful for having morale and anger properly affected by time out of the reality bubble.)

#51194

But... now they heal extremely slowly after reviving so I want to get feedback on whether we need to buff that.

**Pull request merged**